Implement Concat with Strided copy

orausch opened the pull request:
**Description**: 
Adds a `StridedCopy` function that implements a copy from strided tensor to another.

**Motivation and Context**
- This parallelizes the Concat operator, and can also be used in future to parallelize many other data movement operators (e.g. Transpose, Split, etc.).
- This operation is also required for the proposed data layout extensions to ORT. 

**Performance Characteristics**
All benchmarks measure the concatenation of two tensors with feature size 64 and variable batch size.

The PR introduces some overhead at smaller batch sizes: for axis 0, the new PR is at ~most half as slow as the current master with batch sizes less than 2048~ (not anymore, see below). For batch sizes above that, the performance is roughly similar (0.9x-1.05x). 

For multithreaded concatenations, speedup vs master scales slightly less than linearly with the number of cores, as would be expected. Of the (small) tested grid of parameters, the highest measured speedup is ~8x with 12 threads. 
At higher thread counts and high batch sizes, we also measure noticeable (1.5x-6x) speedups vs the parallelized PyTorch implementation.

(Please ping me for the detailed numbers)

orausch commented:
Note: since we did decide to implement a more generalized tensor iterator, this op will likely be better implemented using that abstraction in the future.

orausch commented:
/azp run Linux CPU CI Pipeline,Linux CPU x64 NoContribops CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,MacOS NoContribops CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows WebAssembly CI Pipeline

orausch commented:
/azp run Linux CPU CI Pipeline,Linux CPU x64 NoContribops CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,MacOS NoContribops CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows WebAssembly CI Pipeline

tlh20 commented on onnxruntime/core/providers/cpu/tensor/copy.cc:
```diff
@@ -0,0 +1,151 @@
+                           const TensorShape& copy_shape,
+                           const Tensor& src,
+                           const std::vector<int64_t> src_strides) {
+  ORT_RETURN_IF_NOT(dst.DataType() == src.DataType(), "src and dst types must match");
+  ORT_RETURN_IF_NOT(dst_strides.size() == src_strides.size() && src_strides.size() == copy_shape.NumDimensions(), "src and dst must have same shape");
```
If these tests fail, is it a bug in ORT, or an error in the graph the user provided?  If it's an ORT internal bug then I'd use assert(...) to add the checking rather than exposing an error to the user.

tlh20 commented on onnxruntime/core/providers/cpu/tensor/copy.cc:
```diff
@@ -0,0 +1,151 @@
+                 const std::vector<int64_t>& src_strides) {
+  const auto* src_raw = reinterpret_cast<const uint8_t*>(src);
+  auto* dst_raw = reinterpret_cast<uint8_t*>(dst);
+  const size_t dims = copy_shape.NumDimensions();
```
```suggestion
  const size_t dims = copy_shape.NumDimensions();
  assert(dims >= 1);
```

tlh20 commented on onnxruntime/core/providers/cpu/tensor/copy.cc:
```diff
@@ -0,0 +1,151 @@
+        std::vector<int64_t> current_nd_idx(dims);
+        {
+          int64_t current_index = first;
+          for (size_t dim = dims; dim > 0; dim--) {
+            auto shape_val = copy_shape[dim - 1];
+            // Iterate from dims to 1 so we don't roll over to positive on the bounds check
+            current_nd_idx[dim - 1] = current_index % shape_val;
+            current_index /= shape_val;
+          }
```
```suggestion
            // Iterate from dims to 1 so we don't roll over to positive on the bounds check
          for (size_t dim = dims; dim > 0; dim--) {
            auto shape_val = copy_shape[dim - 1];
            current_nd_idx[dim - 1] = current_index % shape_val;
            current_index /= shape_val;
          }
```

tlh20 reviewed (commented):
Approve re code changes, but could you check with Changming @snnn on any testing needed for the cases where performance degrades (i.e. "the new PR is at most half as slow as the current master with batch sizes less than 2048.").  

I don't have a clear picture on if that is likely to occur in practice.  If it occurs in practice, or we need to be cautious, then can we keep the basic original code, and call into it below a threshold?

Thanks adding the copying microbenchmark test cases for help evaluating that kind of trade-off!

snnn commented:
Thank you. I don't have additional comments.

orausch commented on onnxruntime/core/providers/cpu/tensor/concat.cc:
```diff
@@ -211,50 +212,46 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx,
 
 // This method computes the output tensor for Concat/ConcatFromSequence ops
-Status ConcatBase::ComputeImpl(Prepare& p) const {
+Status ConcatBase::ComputeImpl(Prepare& p, OpKernelContext* ctx) const {
```
All callers have it in scope as `OpKernelContext*`, so I think it makes the most sense to just pass it along.

orausch commented:
> I don't have a clear picture on if that is likely to occur in practice. If it occurs in practice, or we need to be cautious, then can we keep the basic original code, and call into it below a threshold?

I've now added a "fast path" for the 2 dimensional case that is similar to what ORT is was doing before. ORT was previously able to take this path in all concat cases because tensors were always contiguous, but strided copy needs to support more cases than that. 

This path is taken when tensor is <2D. To maximize the cases where it is hit, I've also added contiguous dimension coalescing that reduces the dimensions of the tensors as much as possible.
<details>
<summary> Here is the single threaded profiling after this change (all times in us, median of 500 runs) </summary>

batch_size | feature_size | branch_median | master_median | Speedup Branch vs Master
-- | -- | -- | -- | --
32 | 64 | 3 | 2 | 0.666667
64 | 64 | 3 | 3 | 1
128 | 64 | 4 | 3 | 0.75
256 | 64 | 5 | 5 | 1
512 | 64 | 8 | 8 | 1
1024 | 64 | 22 | 21 | 0.954545
2048 | 64 | 73 | 72 | 0.986301
4096 | 64 | 158.5 | 158 | 0.996845
8192 | 64 | 294 | 286 | 0.972789
16384 | 64 | 614 | 588 | 0.957655
32768 | 64 | 1981.5 | 2061.5 | 1.040373
65536 | 64 | 6308.5 | 6114 | 0.969169

</details>
